### PR TITLE
Purge memprof queue at program exit.

### PR DESCRIPTION
**Pull request by jhjourdan**
When the program exits, there might be some remaining callbacks in the statmemprof queue. This has been discussed in https://github.com/ocaml/ocaml/pull/8920#discussion_r359809307.

@gadmm, you said that you might think that this could not be a good idea because "at_exit is very fragile". Could you expand?

### CODE REVIEW

**Comment by jhjourdan** [comment]
It turns out the test succeeds even without the change in this PR: the reason is that there already is some code running at program exit, that triggers the handling of asynchronous callback by the "normal" way. Still, this change is a good thing to have, because it makes this behavior more robust.

File: stdlib/gc.ml [hunk 1]
```diff
@@ -165,4 +168,6 @@ module Memprof =
       set_ctrl sampling_rate callstack_size minor_alloc_callback
                major_alloc_callback promote_callback minor_dealloc_callback
                major_dealloc_callback
+
+    let () = at_exit (fun () -> if !has_started then stop ())
```
Comment (gasche) [inline stdlib/gc.ml:172]: Can you confirm that unconditionally calling `stop ()` here would be correct (looking at the code of `caml_memprof_set`, I assume that it is an indempotent call that is a no-op if `start ()` was not called before), but the boolean flag is here for performance purposes? (No noticeable computational cost unless Memprof was used.)

I would find it slightly more natural to have a way in the API to ask if memprof is currently sampling (I guess this mean checking `sampling_rate > 0.`), and to stop only in this case. In terms of performance, this is not strictly equivalent, given that your approach will stop even if the program started and then stopped manually.

(Also the implementation would not rely on an extra piece of global mutable state, using only the memprof state.)

**Comment by gasche** [comment]
One aspect of `at_exit` is that it stops at the first exit-handler that raises an exception. Based on our discussion of callback flushing, I suppose that this could also be considered the proper/expected behavior (assuming that we consider at_exit handlers are run asynchronously; this is only true when program termination is itself due to an asynchronous failure).

I wondered if there could be a bad interaction with `Unix.fork` (could the child process' death disrupt the parent's monitoring?) In fact I don't see an issue there. In particular, callbacks defined by the user should not "terminate" (possibly shared) monitoring-related resources, as they have no way to observe that they are called on process-termination rather than on normal deallocation events.

**Comment by gasche** [comment]
To be honest, I have the feeling that we don't understand the issue enough that it's worth working on it for now. I'm happy to keep idly discussing termination situations in #8920 if you find it interesting, but I feel bad diverting your limited time from implementation work that will have more direct impact, in particular working on monitoring of native allocations.

**Comment by gadmm** [comment]
> @gadmm, you said that you might think that this could not be a good idea because "at_exit is very fragile". Could you expand?

Actually I do not think it is such a bad idea, given that the working programmer is not forced to rely on it if they make sure that they properly stop memprof themselves.

From what I gather:

1. The semantics of exceptions arising from at_exit handlers is unclear.
2. The ordering with respect to user-defined at_exit handlers is unclear.

(You can search for issues containing "at_exit" for a scary list of problems, but most of them are solved now.)

@gasche has already mentioned 2. (I did not get what was the conclusion). I checked that it was a least well-defined that `flush_all` is called after memprof callbacks, which is indeed how it should be.

For 1., an exception arising in an at_exit handler can return the control flow to the program, and it can also prevent further at_exit handlers from running.

I think it is fair to ignore the problem of buggy at_exit handlers written by users, and decide that using at_exit will be good enough. However, it would be nice not to raise exceptions from your own at_exit handler, if only to make sure that flush_all from stdlib.ml gets called. Instead, you can ignore it, like `handle_uncaught_exception'` from printexc.ml.

**Review by gadmm — APPROVED** [review]
This PR is nice for convenience for casual Memprof users.

For an express merging, I suggest:
- Catch and ignore the exception arising from `Memprof.stop` inside the `at_exit` handler as is done elsewhere during exit (the only strong request that I have). It is reasonable to expect that the user who is serious about exceptions avoids relying on `at_exit` handlers (as a general rule), for instance here by ensuring Memprof is already stopped.
- I would prefer if `caml_memprof_stop` was made idempotent by replacing the `caml_failwith` with `return Val_unit`, in order to remove `has_started`. The latter is redundant with `started` on the C side, and if it useful for you, then it might also be useful for the user in other circumstances, so making `caml_memprof_stop` idempotent provides them with similar convenience directly. 

@gasche raised other points and after looking at it I do not find other issues to address.

**Comment by xavierleroy** [comment]
Any news?  What needs to be done here?

**Comment by gasche** [comment]
This is a difficult question about a corner case I don't really know how to reason about. Everytime I dive into the details to think about it, I come to a different conclusion. I think we should do one of:
1. Forget about this for now because there is no clear-cut consensus on what the right behavior is. (With apologies to @jhjourdan for the time already invested.)
2. Decide that surely one of the *other* people involved in this discussion (Stephen, Jacques-Henri, Guillaume) is an expert, and their choice should be trusted even in absence of a clear-cut common understanding. Guillaume out-thinks us on these questions, so I would be happy to follow his decision; Jacques-Henri does the work, so he would also be very legitimate to decide. I don't completely understand if they both agree that we should have this PR, or if Jacques-Henri wrote it as a thought experiment in the middle of the discussion (but he's not so convinced himself) and Guillaume approved it because he thinks it moderately improves things (but maybe he isn't fully satisfied with the change either).

**Comment by gadmm** [comment]
As it stands (after #9674 which I assume is going to be merged, in particular https://github.com/ocaml/ocaml/pull/9674/commits/1f544539bce42b07f3b6f99a075827e2032d15dc), this PR no longer works as advertised: there is no queue being purged. Furthermore @jhjourdan explained at the end of #8920 how properly purging the queue is more complicated than one might think with systhreads. It is not clear to me that there will be a meaningful notion of (per-thread? best-effort?) purging after #9674.

Furthermore, the main reasoning for me was that this is a convenience: the careful programmer who made sure that Memprof.stop was properly called would not be affected in any way. With the new behaviour at #9674 where Memprof.stop drops samples, the question is different.

**Comment by gasche** [comment]
Thanks @gadmm for the feedback. My understanding is that we could close this PR -- it seems that people don't care about it so much but that they is real interest in same-thread monitoring. So, with apologies to @jhjourdan, I'm closing. We can reopen if there is renewed interest.

**Comment by jhjourdan** [comment]
@gadmm is right : in #9674, we are doing a best-effort for not dropping samples, but there is not guarantee. So the current PR is not relevant.

**Pull request closed without merging**